Vita port

isage opened the pull request:
That's a quick'n'dirty PSVita port and static-linking fix.
This probably shouldn't be merged as-is, and instead used as a base for proper static-linking support by someone, who's familiar with libretro/retroarch (because i'm not)

isage committed:
Use vita funcs

isage committed:
Fix static linking

inactive123 commented:
Hi @schellingb , 

how should we approach this PR? Merge it as-is, or do you want some alterations to be made to it? Let me know and as far as I'm concerned you can review and supervise this PR.

schellingb commented:
This looks perfectly fine to me as-is. I guess in addition to that we'd need to enable the platform in .gitlab-ci.yml to get it also building on the buildbot.
I don't have a Vita (or any static linking platform for that matter) to test but I assume the core is working for you with these changes @isage? Someone on Twitter mentioned to me today that the Vita port is unable to load ZIP files, is that still the case here? And if so, do you have any idea why that wouldn't work?

isage commented:
Honestly, no idea why zip doesn't/wouldn't work.
One user reported that he can't load zip or exe (same one as on Twitter, i guess), another that he tested with couple of games and everything is fine.
I'm trying to get details from him rn

isage commented:
```
sarcastic_cat — Today at 5:46 PM
Does zip loading work?

EmuMusicFan — Today at 5:47 PM
Not tested yet, but SoundFont support works well with DOOM.
Not sure how performance demanding this SoundFont MIDI synthesizer is, but I know Ad Lib is still quite performance hungry.
I will test MT-32 support later.

sarcastic_cat — Today at 5:50 PM
Can you check please when you have time? zip and dosz?

EmuMusicFan — Today at 5:51 PM
Just loaded a zip file without dos contents, and it works, just tell me No executable file found.
Now let me pack a zip with DOOM in.

sarcastic_cat — Today at 5:51 PM
And if you have github account - post this in the pullrequest linked above

EmuMusicFan — Today at 5:52 PM
Append a result to your PR? Got it!

sarcastic_cat — Today at 5:52 PM
yeah. thanks!

EmuMusicFan — Today at 5:58 PM
Welcome, but it is late in my time zone, so I think I could post it later.
Oh, DOOM in zip does not work.
Listing files in the package is OK, but it could not executable doom.exe, and setup is not OK either. Even my launcher bat doesn't work.

sarcastic_cat — Today at 6:04 PM
what if you rename it to dosz?

EmuMusicFan — Today at 6:05 PM
Wait a moment. Something interesting happened.
It works with a "store" zip!
I used default zip compression earlier, and it didn't work.
Just now I repacked the zip with "store" and it works.
```

schellingb commented:
Huh that's funny. Good catch with figuring out that it works without compression.
The decompression code should all be very portable and certainly works perfectly fine on 32 bit Raspberry PI (ARMv7)... I assume there's no compiler warnings or anything that would indicate something wrong in drive_zip.cpp?

isage commented:
Nope.

isage commented:
Well, i guess compression can be debugged and dealt with later, after merging this and bringing core to vita, so more users can test

schellingb commented:
Loading ZIP files is the main feature of this core though... Is there any way I can try to test this without an actual device? Can I do a static build on Linux to see if that's somehow the reason why it breaks? Or do you know if there is a Vita emulator that can run RetroArch? No need to be fast, just enough to run this core and replicate the problem.

isage commented:
Oof. There's only one vita emulator - Vita3k, as to whether it can run retroarch - i don't know. Maybe @frangarcj can help

frangarcj commented:
Vita3k could do the trick. I will try it too

schellingb commented:
@isage Can you upload a .vpk file with this core for reference?
I built RetroArch for vita and had a bunch of issues. I needed to manually compile texture_tint_v_gxp.c and add -lSceAppUtil_stub to the linker, then I added retroarch_vita.elf to retroarchvita_salamander.vpk to install it. When launching it with vita3k, it boot loops, opening and closing the emulator window until I press ctrl-c. Though I don't have much confidence in my vpk.

Edit: I got further. I ran `make -f Makefile.griffin platform=vita` again and it ended up outputting a bigger eboot.bin that I placed into retroarchvita_salamander.vpk and launching that with vita3k brings up RetroArch. But it crashes when selecting "Start Core" and even after enabling logging nothing useful is written to retroarch.log (it's not empty so logging works at least). It shows "1.9.3 - DOSBox-pure (0.13)" in the lower left corner of the RetroArch menu, so it seems to be close.

schellingb commented:
Well... After manually writing an entry to content_history.lpl it just worked. With a ZIP file with compressed files in it. So uhhh... I'd say, let's merge this :D
![image](https://user-images.githubusercontent.com/14200249/118393216-e937a180-b678-11eb-92ac-858557290dde.png)

Invictaz commented:
@schellingb it doesn't work on real hardware (the zip part) it crashes. Also .dosz crashes.

The second problem is that the automatic controller mapping from bigjim doesn't recognize and work with the Vita buttons. Thus every game needs it's own map file. It's really frustrating to remap each game. 

I can send you the error message of the build @isage gave me. The eboot.bin is a renamed .self

Since there is no L3 button you have to remap the on screen keyboard for each game. 

Plus it doesn't detect the left or right stick as mouse.

For the rest, only non compressed zips (store zip) worked for me.

schellingb commented:
@Invictaz Thanks for the detailed report!

I still think there is a chance that the problem with compressed files is not real hardware vs. emulator but the compilation. And thus if we get the official libretro buildbot to build there still is a possibility that it ends up actually working. The actual decompression code is pretty much pure portable C code and unless the real hardware runs out of memory (which should follow a full core crash) I have a hard time thinking of actual problems besides issues during compilation. I'm willing to give the buildbot a shot.

The automatic controller mappings of DOSBox Pure are mapping to the RetroPad and not a specific type of gamepad. If there are Vita specific problems, they'd need to be fixed in RetroArch itself. The controller mapping can be set core wide so only needs to be done once. Also doesn't Vita have multiple touch surfaces that could be used as L3?

> I can send you the error message of the build @isage gave me. The eboot.bin is a renamed .self

Is this error message about buttons or about file loading? Either way, it would be helpful to post any warnings/errors you get here so I can at least think about them :-) Now that I can build and run some form of a Vita build I feel I can be more useful.

Invictaz commented:
@schellingb here is the error log. Tried with the following .dosz (Duke Nukem) 

(the .dosz is inside of a Zipfile because of Github rules). Zo unpack the .dosz and try it. 

[Duke Nukem -  Episode 1 - Shrapnel City (1991).zip](https://github.com/libretro/dosbox-pure/files/6509166/Duke.Nukem.-.Episode.1.-.Shrapnel.City.1991.zip)

Errorlog is here

[psp2core-1621431549-0x00000b2fb7-dosbox_pure_libretro.self.zip](https://github.com/libretro/dosbox-pure/files/6509150/psp2core-1621431549-0x00000b2fb7-dosbox_pure_libretro.self.zip)

isage commented:
> @schellingb here is the error log. Tried with the following .dosz (Duke Nukem)
> 
> (the .dosz is inside of a Zipfile because of Github rules). Zo unpack the .dosz and try it.
> 
> [Duke Nukem - Episode 1 - Shrapnel City (1991).zip](https://github.com/libretro/dosbox-pure/files/6509166/Duke.Nukem.-.Episode.1.-.Shrapnel.City.1991.zip)
> 
> Errorlog is here
> 
> [psp2core-1621431549-0x00000b2fb7-dosbox_pure_libretro.self.zip](https://github.com/libretro/dosbox-pure/files/6509150/psp2core-1621431549-0x00000b2fb7-dosbox_pure_libretro.self.zip)

That's not error log, though. That's crash coredump

Invictaz commented:
I don't have any better log. Maybe a debug build will show what is wrong. 
It's quite strange as @schellingb tried it on Vita3k. Not sure if he also tried .dosz

isage commented:
Decoded coredump from debug version
```
=== THREADS ===
    RETROVITA
        ID: 0x40010003
        Stop reason: 0x0 (No reason)
        Status: 0x8 (Waiting)
        PC: 0xe000bee4 (SceLibKernel@1 + 0x6714 => b'menu_entries_append_enum at /media/2TB3/libretro-super/retroarch/retroarch.c:3441')
    SceCommonDialogWorker
        ID: 0x400100c3
        Stop reason: 0x0 (No reason)
        Status: 0x8 (Waiting)
        PC: 0xe000bad4 (SceLibKernel@1 + 0x6304 => b'menu_entry_get at /media/2TB3/libretro-super/retroarch/retroarch.c:2836')
    
        ID: 0x4001012d
        Stop reason: 0x0 (No reason)
        Status: 0x8 (Waiting)
        PC: 0xe000bad4 (SceLibKernel@1 + 0x6304 => b'menu_entry_get at /media/2TB3/libretro-super/retroarch/retroarch.c:2836')
    
        ID: 0x40010139
        Stop reason: 0x30004 (Data abort exception)
        Status: 0x1 (Running)
        PC: 0x81198d92 (retroarch_vita.elf@1 + 0x17fd92 => b'Zip_DeflateMemoryUnpacker::Zip_DeflateMemoryUnpacker(Zip_Archive&, Zip_File const&) at /media/2TB3/libretro-super/libretro-dosbox_pure/src/dos/drive_zip.cpp:1069')
    SceGxmDisplayQueue
        ID: 0x4001014b
        Stop reason: 0x0 (No reason)
        Status: 0x8 (Waiting)
        PC: 0x812faea8 (retroarch_vita.elf@1 + 0x2e1ea8 => b'sceDisplayWaitVblankStart at ??:?')
    
        ID: 0x40010183
        Stop reason: 0x0 (No reason)
        Status: 0x8 (Waiting)
        PC: 0x812fb518 (retroarch_vita.elf@1 + 0x2e2518 => b'sceAudioOutOutput at ??:?')

=== THREAD "" <0x40010139> CRASHED (Data abort exception) ===

DISASSEMBLY AROUND PC: 0x81198d92 (ARM):

8117fd82 <_ZN8zipDriveC1EP8DOS_Fileb+0x3a>:
};

zipDrive::zipDrive(DOS_File* zip, bool enter_solo_root_dir) : impl(new zipDriveImpl(zip, enter_solo_root_dir))
{
        label.SetLabel("ZIP", false, true);
}
8117fd82:       815d            .short  0x815d
8117fd84:       815ab460        .word   0x815ab460

8117fd88 <_ZN25Zip_DeflateMemoryUnpackerC1ER11Zip_ArchiveRK8Zip_File>:
        Zip_DeflateMemoryUnpacker(Zip_Archive& archive, const Zip_File& f)
8117fd88:       e92d 4ff0       stmdb   sp!, {r4, r5, r6, r7, r8, r9, sl, fp, lr}
8117fd8c:       f5ad 4dd6       sub.w   sp, sp, #27392  ; 0x6b00
8117fd90:       b085            sub     sp, #20
!!!             9009            str     r0, [sp, #36]   ; 0x24 !!!
8117fd94:       460f            mov     r7, r1
8117fd96:       4614            mov     r4, r2
8117fd98:       9207            str     r2, [sp, #28]
        pointer _M_start;
        pointer _M_finish;
        pointer _M_end_of_storage;

        _Vector_impl_data() _GLIBCXX_NOEXCEPT
        : _M_start(), _M_finish(), _M_end_of_storage()
8117fd9a:       2300            movs    r3, #0
8117fd9c:       6043            str     r3, [r0, #4]
8117fd9e:       6083            str     r3, [r0, #8]
8117fda0:       60c3            str     r3, [r0, #12]


DISASSEMBLY AROUND LR: 0x8119917a (Thumb):

8118016a <_ZN10Zip_HandleC1ER11Zip_ArchiveP8Zip_FilejPKc+0x15e>:
                        if (!_src->uncomp_size) _src->unpacker = nullptr;
                        else if (_src->method == ZIP_Unpacker::METHOD_DEFLATED)
                        {
                                enum { MINIMAL_SIZE = (sizeof(Zip_DeflateUnpacker) + sizeof(Zip_DeflateUnpacker::SeekCursor)) };
                                if (_src->uncomp_size > MINIMAL_SIZE) _src->unpacker = new Zip_DeflateUnpacker(archive, *_src);
                                else                                  _src->unpacker = new Zip_DeflateMemoryUnpacker(archive, *_src);
8118016a:       2010            movs    r0, #16
8118016c:       f0e4 fb50       bl      81264810 <_Znwj>
81180170:       4682            mov     sl, r0
81180172:       462a            mov     r2, r5
81180174:       4631            mov     r1, r6
81180176:       f7ff fe07       bl      8117fd88 <_ZN25Zip_DeflateMemoryUnpackerC1ER11Zip_ArchiveRK8Zip_File>
!!!             f8c5 a028       str.w   sl, [r5, #40]   ; 0x28 !!!
8118017e:       e797            b.n     811800b0 <_ZN10Zip_HandleC1ER11Zip_ArchiveP8Zip_FilejPKc+0xa4>
                        }
                        else if (_src->method == ZIP_Unpacker::METHOD_STORED)   _src->unpacker = new Zip_StoredUnpacker(archive);
81180180:       2008            movs    r0, #8
81180182:       f0e4 fb45       bl      81264810 <_Znwj>
        Zip_StoredUnpacker(Zip_Archive& _archive) : archive(_archive) { }
81180186:       495c            ldr     r1, [pc, #368]  ; (811802f8 <_ZN10Zip_HandleC1ER11Zip_ArchiveP8Zip_FilejPKc+0x2ec>)
81180188:       6001            str     r1, [r0, #0]

REGISTERS:
    R0: 0x86e6fdb8
    R1: 0x8646a168
    R2: 0x8646e450
    R3: 0x0
    R4: 0x86e76178
    R5: 0x8646e450
    R6: 0x8646a168
    R7: 0x816bca1c
    R8: 0xbc00
    R9: 0x20
    R10: 0x86e6fdb8
    R11: 0x16b28
    R12: 0x0
    SP: 0x816b5d50
    PC: 0x81198d92 (retroarch_vita.elf@1 + 0x17fd92)
    LR: 0x8119917b (retroarch_vita.elf@1 + 0x18017b)

STACK CONTENTS AROUND SP:
```

isage commented:
Forcefully replacing Zip_DeflateMemoryUnpacker with Zip_DeflateUnpacker on vita fixes crash.

schellingb commented:
Awesome report, thanks @Invictaz and @isage 
So the constructor of Zip_DeflateMemoryUnpacker uses 27kb of stack memory which is large but not obscenely large. It looks like Vita has a very small stack size for threads (maybe as low as 16kb?) so now with 800a385 I correctly set a default size for the stack of the one main emulation thread.

Can we quickly confirm that this indeed fixes it on real hardware and then proceed with the merge?

Btw. I noticed in Vita3k after 2 or 3 minutes Vita3k just freezes. Both DOOM and Duke Nukem, as well as just sitting at the command prompt. Audio stops and the image stops moving, nothing is logged to the Vita3k console. Not sure if that's a Vita3k issue or if there are further problems.

Invictaz commented:
@schellingb well @isage reported on Discord some crashes and freezes as well.

As for the zip crashes, he described the Zip_DeflateUnpacker fix as a workaround though:
"There are two decompression classes selected based on decompressed file size. It uses memory one for file sizes below some threshold. I removed that check to force another decompression class. That's not a proper fix, though, just a way to pinpoint problem".

So maybe instead of decompressing zip to memory, it now uses write cycles?
That would be wearing down the memorycard, especially on larger zipfiles. As the Vita memorycards are very very expensive we should be careful doing that.

schellingb commented:
My fix correctly addresses the core problem of a small stack size. The ZIP code did not need to be touched at all.
I could have moved data from the stack to the heap but similar to Zip_DeflateMemoryUnpacker the code of GUS, MT32 and some internal DOSBox programs have large variables on the stack. They'd all need to be moved out and there still is a high chance of the code hitting the tiny stack limit. So fixing the size for all platforms was certainly needed and hopefully increases consistent behavior across platforms.

Invictaz commented:
@schellingb Not to be running ahead but porting the core to the Sony PSP could introduce an even smaller stack size?

There is now a new Dosbox for the PSP https://github.com/PSP-Archive/dosbox_psp
it hopefully leads to Dosbox Pure on that system.

I saw also a port to the GCW0. All these smaller game consoles might need a proper fix?
https://github.com/schellingb/dosbox-pure/issues/79

schellingb commented:
The base model of the PSP has 32MB RAM, I think that's maybe a bit too low for both RetroArch + DOSBox Pure. The linked DOSBox PSP port by default allocates 12 MB DOS memory (leaving mere 4MB for the video emulation, the entire emulation stack and the UI). Maybe it'd be possible to get it running on PSP slim (which has 64MB)?

For all these platforms that I don't own or can't build myself for I 100% rely on contributions. I tried enabling building for other consoles but it never quite worked out. I think with Vita now potentially working, others can be reasonably tackled as well. As long as there is someone who can get a crash dump analyzed :-)

Invictaz commented:
The port I linked is a standalone Dosbox. Does Dosbox Pure need Retroarch as a requirement?
I know that Retroarch for PSP recently received a lot of cores for older systems and was optimized a lot with a newer toolchain, so that the framerate of cores was improved vastly. 

What Vita needs is proper handling of the control remapping. Otherwise making a map file for each game is a pain. 

Also one of the features of Dosbox Pure is that savegames are saved in separate zipfiles. These are not properly generated now on the Vita. I checked it. 

This leaves only the .SAV and .STATE files which are build into Retroarch. So saving a dosgame in it's own menu's does nothing for now.

isage committed:
Adjust vita build flags

isage commented:
> Awesome report, thanks @Invictaz and @isage
> So the constructor of Zip_DeflateMemoryUnpacker uses 27kb of stack memory which is large but not obscenely large. It looks like Vita has a very small stack size for threads (maybe as low as 16kb?) so now with [800a385](https://github.com/libretro/dosbox-pure/commit/800a385969de8d0608b7dda13c580d47961c4f74) I correctly set a default size for the stack of the one main emulation thread.
> 
> Can we quickly confirm that this indeed fixes it on real hardware and then proceed with the merge?

Can confirm that it works. I've rebased my branch and added additional commit adjusting build flags on vita, as it's armv7 and vitasdk uses gcc 10.2, so building with -O2 was causing dynarec to fail.

Invictaz commented:
@schellingb There are several things wrong. If you exit a game and it goes back to DOS, it says in the particular dos font "press any key to exit". If you do that, Retroarch quits to the Vita Livearea (aka the whole app quits) 

Plus the automatic controller mappings are still not working well. Are they mapped to a joystick or to a keyboard?

It seems that with the last merge, Dosbox Pure is now in the nightly builds and should load all .dosz and compressed zipfiles. However I need to check it more.
